src/server/pfs/server/driver_new_storage.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/tar"
+	txnenv "github.com/pachyderm/pachyderm/src/server/pkg/transactionenv"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"golang.org/x/net/context"
+)
+
+const (
+	compacted = "compacted"
+)
+
+func init() {
+	// This is kind of weird, but I think it might be necessary.
+	// Since we build our protos with gogoproto, we need to register
+	// the types that can be embedded in an Any type with the gogoproto
+	// proto package because building the protos just registers them with the
+	// normal proto package.
+	proto.RegisterType((*pfs.Merge)(nil), "pfs.Merge")
+	proto.RegisterType((*pfs.Shard)(nil), "pfs.Shard")
+}
+
+var (
+	// (bryce) need to expose as configuration.
+	memThreshold int64 = 1024 * chunk.MB
+)
+
+func (d *driver) startCommitNewStorageLayer(txnCtx *txnenv.TransactionContext, id string, parent *pfs.Commit, branch string, provenance []*pfs.CommitProvenance, description string) (*pfs.Commit, error) {
+	commit, err := d.startCommit(txnCtx, id, parent, branch, provenance, description)
+	if err != nil {
+		return nil, err
+	}
+	d.fs = d.storage.New(context.Background(), commit.ID, fileset.WithMemThreshold(memThreshold))
+	return commit, nil
+}
+
+func (d *driver) finishCommitNewStorageLayer(txnCtx *txnenv.TransactionContext, commit *pfs.Commit, description string) (retErr error) {
+	if err := d.checkIsAuthorizedInTransaction(txnCtx, commit.Repo, auth.Scope_WRITER); err != nil {
+		return err
+	}
+	commitInfo, err := d.resolveCommit(txnCtx.Stm, commit)
+	if err != nil {
+		return err
+	}
+	if commitInfo.Finished != nil {
+		return pfsserver.ErrCommitFinished{commit}
+	}
+	if description != "" {
+		commitInfo.Description = description
+	}
+	// Close in-memory file set (serializes in-memory part).
+	if err := d.fs.Close(); err != nil {
+		return err
+	}
+	defer func() {
+		if retErr == nil {
+			d.fs = nil
+		}
+	}()
+	// Merge the commit with its parent.
+	// (bryce) this should be where the level based compaction scheme is used.
+	// right now we are eagerly compacting at every commit.
+	fileSets := []string{commitInfo.Commit.ID}
+	parentCommit := commitInfo.ParentCommit
+	if parentCommit != nil {
+		// (bryce) how to handle parent commit that is not closed?
+		parentCommitInfo, err := d.resolveCommit(txnCtx.Stm, parentCommit)
+		if err != nil {
+			return err
+		}
+		fileSets = append(fileSets, path.Join(parentCommitInfo.Commit.ID, compacted))
+	}
+	if err := d.merge(context.Background(), path.Join(commitInfo.Commit.ID, compacted), fileSets); err != nil {
+		return err
+	}
+	// (bryce) need size.
+	commitInfo.SizeBytes = uint64(0)
+	commitInfo.Finished = now()
+	return d.writeFinishedCommit(txnCtx.Stm, commit, commitInfo)
+}
+
+// (bryce) should have a context going through here.
```

Comment:
`pachClient` has a context in it for you to use

Reply:
See above.

Reply:
Sorry not sure where above I'm supposed to be looking.

Reply:
The immediately prior review comment thread.

Reply:
Ahh, I get it, we'll just continue the chat there.